### PR TITLE
Implement PushTraces

### PR DESCRIPTION
**Pull request by srikanthccv**
Part 3 of breakdown https://github.com/SigNoz/signoz-otel-collector/pull/479

- periodically fetch the count from DB to decided if a key should be skipped
- maintain a local tracker to skip a key
- if it passes both stages, then check in fingerprint cache to see if it has been inserted already
- if not, then insert

### CODE REVIEW

File: exporter/metadataexporter/exporter.go [hunk 1]
```diff
@@ -19,21 +34,385 @@ type tagValueCountFromDB struct {
...
+			logger: e.set.Logger,
+			conn:   e.conn,
+			query: `SELECT tag_key, tag_data_type, countDistinct(string_value) as string_value_count, countDistinct(number_value) as number_value_count
+						 FROM signoz_traces.distributed_tag_attributes_v2
+						 GROUP BY tag_key, tag_data_type
+						 ORDER BY number_value_count DESC, string_value_count DESC, tag_key
+						 LIMIT 1 BY tag_key, tag_data_type`,
+			storeFunc:  e.storeTracesTagValues,
+			signalName: pipeline.SignalTraces.String(),
+			interval:   15 * time.Minute,
```
Comment (nityanandagohain) [inline exporter/metadataexporter/exporter.go:125]: this was to be 10 mins ?

File: exporter/metadataexporter/exporter.go [hunk 2]
```diff
@@ -19,21 +34,385 @@ type tagValueCountFromDB struct {
...
+
+	e.fingerprintCache.Set(cacheKey, true, ttlcache.DefaultTTL)
+	return false, nil
+}
+
 func (e *metadataExporter) PushTraces(ctx context.Context, td ptrace.Traces) error {
+	stmt, err := e.conn.PrepareBatch(ctx, "INSERT INTO signoz_metadata.distributed_attributes_metadata", driver.WithReleaseConnection())
+	if err != nil {
+		return err
+	}
```
Comment (nityanandagohain) [inline exporter/metadataexporter/exporter.go:352]: add defer functions which aborts the statement ?

File: exporter/metadataexporter/exporter.go [hunk 3]
```diff
@@ -19,21 +34,385 @@ type tagValueCountFromDB struct {
...
+					e.set.Logger.Error("failed to write to stmt", zap.Error(err))
+				}
+				if skipped {
+					skippedSpans++
+				}
+			}
+		}
+	}
+
+	e.set.Logger.Info("pushed traces attributes", zap.Int("total_spans", totalSpans), zap.Int("skipped_spans", skippedSpans))
```
Comment (nityanandagohain) [inline exporter/metadataexporter/exporter.go:411]: required ? can be debug

File: exporter/metadataexporter/exporter.go [hunk 4]
```diff
@@ -19,21 +34,385 @@ type tagValueCountFromDB struct {
...
+				if len(skippedFromDB) > 0 {
+					e.set.Logger.Debug("skipped attributes from DB (traces)", zap.Strings("keys", skippedFromDB))
+				}
+
+				flattenedSpanAttrs := flatten.FlattenJSON(spanAttrs, "")
+				filteredSpanAttrs := e.filterAttrs(ctx, flattenedSpanAttrs, pipeline.SignalTraces.String())
+				spanFingerprint := fingerprint.FingerprintHash(filteredSpanAttrs)
+
+				unixMilli := span.StartTimestamp().AsTime().UnixMilli()
+				roundedUnixMilli := (unixMilli / sixHoursInMs) * sixHoursInMs
```
Comment (nityanandagohain) [inline exporter/metadataexporter/exporter.go:398]: roundedSixHrsUnixMilli maybe

**Review by nityanandagohain — COMMENTED** [review]
Minor comments

File: exporter/metadataexporter/exporter.go (hunk 1 above)
Comment (srikanthccv) [reply exporter/metadataexporter/exporter.go:125]: It has been 15 minutes in the test environment, we can probably revisit it later

File: exporter/metadataexporter/exporter.go (hunk 2 above)
Comment (srikanthccv) [reply exporter/metadataexporter/exporter.go:352]: I read about it and don't really understand why it's required. Please help me understand why we explicitly need to abort?

Comment (nityanandagohain) [reply exporter/metadataexporter/exporter.go:358]: I tried checking it and didn't find anything concrete, though some of their test does contain the abort statement for batches.

**Review by nityanandagohain — APPROVED** [review]

**Pull request merged**